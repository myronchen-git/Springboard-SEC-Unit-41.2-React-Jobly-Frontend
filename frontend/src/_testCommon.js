const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
