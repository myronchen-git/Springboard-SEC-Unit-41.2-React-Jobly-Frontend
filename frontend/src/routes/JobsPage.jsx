import { useEffect, useState } from 'react';

import JoblyApi from '../../../api';
import JobCard from '../components/JobCard';
import SearchBar from '../components/SearchBar';

// ==================================================

/**
 * Lists all jobs.  Allows the list to be filtered.
 */
function JobsPage() {
  const [jobs, setJobs] = useState([]);
  const [errorMessage, setErrorMessage] = useState();

  useEffect(() => {
    retrieveJobs();
  }, []);

  async function retrieveJobs(filters = {}) {
    try {
      setJobs(await JoblyApi.getJobs(filters));
    } catch (err) {
      setErrorMessage(err[0]);
    }
  }

  return (
    <main className="JobsPage">
      <SearchBar retrieveItems={retrieveJobs} filterName="title" />
      {errorMessage ? (
        <div className="JobsPage__error">
          <p>{errorMessage}</p>
        </div>
      ) : (
        <ul>
          {jobs.map((job) => (
            <li key={job.id}>
              <JobCard job={job} isApplied={false} />
            </li>
          ))}
        </ul>
      )}
    </main>
  );
}

// ==================================================

export default JobsPage;
