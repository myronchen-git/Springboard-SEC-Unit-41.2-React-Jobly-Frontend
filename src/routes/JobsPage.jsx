import { useContext, useEffect, useState } from 'react';

import JoblyApi from '../api.js';
import JobCard from '../components/JobCard.jsx';
import SearchBar from '../components/SearchBar.jsx';
import { UserContext } from '../contexts.jsx';

// ==================================================

/**
 * Lists all jobs.  Allows the list to be filtered.
 */
function JobsPage() {
  const [jobs, setJobs] = useState([]);
  const [errorMessage, setErrorMessage] = useState();
  const { applications } = useContext(UserContext);

  useEffect(() => {
    retrieveJobs();
  }, []);

  async function retrieveJobs(filters = {}) {
    try {
      setJobs(await JoblyApi.getJobs(filters));
    } catch (err) {
      setErrorMessage(err[0]);
    }
  }

  return (
    <main className="JobsPage">
      <SearchBar retrieveItems={retrieveJobs} filterName="title" />
      {errorMessage ? (
        <div className="JobsPage__error">
          <p>{errorMessage}</p>
        </div>
      ) : (
        <ul>
          {jobs.map((job) => (
            <li key={job.id}>
              <JobCard job={job} isApplied={applications.includes(job.id)} />
            </li>
          ))}
        </ul>
      )}
    </main>
  );
}

// ==================================================

export default JobsPage;
