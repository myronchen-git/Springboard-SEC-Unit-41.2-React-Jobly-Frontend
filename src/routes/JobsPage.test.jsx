import { act, fireEvent, render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { jobs } from '../_testCommon.js';
import JoblyApi from '../api.js';
import { UserContext } from '../contexts.jsx';
import JobsPage from './JobsPage.jsx';

// ==================================================

vi.mock(import('../api.js'), () => {
  const MockJoblyApi = vi.fn();
  MockJoblyApi.getJobs = vi.fn();

  return {
    default: MockJoblyApi,
  };
});

// ==================================================

describe('JobsPage', () => {
  const applications = Object.freeze(['2', '3']);
  const mockApplyToJob = vi.fn();

  beforeEach(() => {
    mockApplyToJob.mockReset();

    JoblyApi.getJobs.mockImplementation((filters) => {
      let jobsToReturn;

      if (Object.keys(filters).length) {
        // If there are filters.

        const jobTitle = filters.title;

        jobsToReturn = jobs.reduce((arr, job) => {
          if (job.title.toLowerCase().includes(jobTitle.toLowerCase())) {
            arr.push({ ...job });
          }

          return arr;
        }, []);
      } else {
        // If there are no filters.

        jobsToReturn = structuredClone(jobs);
      }

      return Promise.resolve(jobsToReturn);
    });
  });

  it('renders.', async () => {
    await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <JobsPage />
        </UserContext.Provider>
      )
    );
  });

  it('matches snapshot.', async () => {
    const { asFragment } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <JobsPage />
        </UserContext.Provider>
      )
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays jobs and search bar.', async () => {
    // Act
    const { getByPlaceholderText, getByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <JobsPage />
        </UserContext.Provider>
      )
    );

    // Assert
    jobs.forEach((job) => {
      expect(getByText(job.title)).toBeVisible();
    });
    expect(getByPlaceholderText('Enter search term...')).toBeVisible();
  });

  it('filters the list of jobs.', async () => {
    // Arrange
    const { getByPlaceholderText, getByText, queryByText } = await act(
      async () =>
        render(
          <UserContext.Provider
            value={{ applications, applyToJob: mockApplyToJob }}
          >
            <JobsPage />
          </UserContext.Provider>
        )
    );

    const searchInput = getByPlaceholderText('Enter search term...');
    const submitBtn = getByText('Search');

    // Act
    fireEvent.change(searchInput, { target: { value: jobs[0].title } });
    await act(async () => fireEvent.click(submitBtn));

    // Assert
    expect(getByText(jobs[0].companyName)).toBeVisible();

    jobs.forEach((job, i) => {
      if (i !== 0) {
        expect(queryByText(job.title)).not.toBeInTheDocument();
      }
    });
  });

  it('displays the correct apply button status.', async () => {
    // Act
    const { getAllByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <JobsPage />
        </UserContext.Provider>
      )
    );

    // Assert
    // Fragile!
    const applybtns = getAllByText('Appl', { exact: false });

    expect(applybtns[0]).toHaveTextContent('Apply');
    expect(applybtns[1]).toHaveTextContent('Applied');
  });

  it('allows applying to a job.', async () => {
    // Arrange
    const { queryAllByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <JobsPage />
        </UserContext.Provider>
      )
    );

    const applyBtns = queryAllByText('Apply');

    // Act
    await act(async () => fireEvent.click(applyBtns[0]));

    // Assert
    expect(mockApplyToJob).toHaveBeenCalledOnce();
    expect(mockApplyToJob).toHaveBeenCalledWith(jobs[0].id);
  });
});
