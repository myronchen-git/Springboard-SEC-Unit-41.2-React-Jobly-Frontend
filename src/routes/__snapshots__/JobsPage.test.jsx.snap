// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JobsPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="JobsPage"
  >
    bound HTMLFormElement {
      "0": <input
        class="form-control"
        name="title"
        placeholder="Enter search term..."
        type="search"
        value=""
      />,
      "1": <button
        class="btn btn-secondary"
        type="submit"
      >
        Search
      </button>,
    }
    <ul
      class="JobsPage__companies-list list-unstyled"
    >
      <li>
        <div
          class="JobCard card"
        >
          <div
            class="JobCard__info card-body"
          >
            <h4
              class="card-title"
            >
              Job 1
            </h4>
            <div
              class="card-subtitle"
            >
              Company 1
            </div>
            <p
              class="card-text"
            >
              Salary: $1
            </p>
            <p
              class="card-text"
            >
              Equity: 0.1
            </p>
          </div>
          bound HTMLFormElement {
            "0": <button
              class="btn btn-secondary"
              type="button"
            >
              Apply
            </button>,
          }
        </div>
      </li>
      <li>
        <div
          class="JobCard card"
        >
          <div
            class="JobCard__info card-body"
          >
            <h4
              class="card-title"
            >
              Job 2
            </h4>
            <div
              class="card-subtitle"
            >
              Company 1
            </div>
            <p
              class="card-text"
            >
              Salary: $2
            </p>
            <p
              class="card-text"
            >
              Equity: 0.2
            </p>
          </div>
          bound HTMLFormElement {
            "0": <button
              class="btn btn-secondary disabled"
              disabled=""
            >
              Applied
            </button>,
          }
        </div>
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
